new doc title and revised intro (#638)
input-output-hk/plutus-apps: doc/index.rst

> Plutus Tools SDK User Guide

Reviewer: Hmm, do we need to duplicate the title here? Can we get sphinx to print it here or something?

Author: Good question. When I removed the H1 heading here, it also impacted the title displaying above. I would think there would be a way to avoid the duplication. I'll look into that.

Author: In the current versions of both RTD instances, the title is duplicated. I would like to find a way to not duplicate it, but just want to note that this isn't a new issue. But I'll look into this to see if we can improve this behavior.

===============================

Intended Audience
-------------------------------

External Developers

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: Isn't it redundant to have a separate section for "External developers" when this doc is specifically for external developers?

Reviewer: Also, this is pretty weird to use in an users guide! From the users guide perspective "normal" developers are "external", we're weird for being "internal"!

Author: We could probably go without the "External Developers" heading. My goal is that if a reader who knew nothing about any of this came to this page for the first time, I want them to understand what this documentation is, who it's for, why they may or may not care about it. Easy to understand and clear.

> This user guide is intended for external developers who are authoring distributed 

Reviewer: ```suggestion
This user guide is intended for developers who are authoring distributed 
```

Author: Suggestion incorporated. 👍

> applications ("DApps") that implement smart contracts on the Cardano blockchain. 

Author: Suggestion incorporated. 👍

.. note::
   If you are a developer who wants to contribute to the Plutus Tools SDK project, 
   please refer to documentation residing in the `Plutus Tools SDK repository <https://github.com/
   input-output-hk/plutus-apps>`_.

Plutus Tools SDK
===============================

The Plutus Tools SDK is a collection of off-chain infrastructure resources built 
for external developers. The `Plutus Tools SDK repository <https://github.com/
input-output-hk/plutus-apps>`_ supports the underlying resources that developers 
need who are writing full applications using Plutus in Haskell, including off-chain 
code. The term “off-chain code” refers to the part of a contract application’s code 

> which runs off the chain. 

Reviewer 2: For more precision, how about "... which runs outside the blockchain"?

Author: Yes, I think that's a helpful description.

> code. The term “off-chain code” refers to the part of a contract application’s code 
> which runs off the chain. 

Reviewer 3: ```suggestion
code. The term “off-chain code” refers to the part of a contract application’s code 
which responds to events happening on the blockchain, usually by producing transaction. 
```
I am not sure about the correct, but the original text sounds a bit like a circular definition.

Author: This adds some useful perspective on what off-chain code is doing.

Author: @koslambrou @raduom In light of your suggestions, this is what I'm going with right now: 

> The term “off-chain code” refers to the part of a contract application’s code 
> which runs outside of the blockchain. Off-chain code responds to events happening 
> on the blockchain, usually by producing transactions.

The `Plutus Tools SDK repository <https://github.com/input-output-hk/plutus-apps>`_ 
contains packages such as: 

* the `Plutus Application Backend (PAB) <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-pab>`_, 
  an off-chain application for managing the state of Plutus contract instances;

* the `chain-index <https://github.com/input-output-hk/plutus-apps/tree/main/plutus-chain-index>`_, 
  a lightweight, customizable chain follower application and library for DApp 
  developers who need to index and query the Cardano blockchain; 

> * a variety of other Plutus packages. 

Reviewer 3: I think this says too little. Some things that I would like to know:
* What packages? Are they packages or Haskell modules?
* What are they used for?

I think just 1-2 examples would do, but without those this seems quite superfluous as it does not provide any new information.

Author: @raduom I agree with your comment. I would like to include more substantial information here, but I'm unclear about what is available today from Tools that is ready for developers to use. What else could we add here, and what else might we say about what is available? 

@michaelpj Excellent point. I would like to have a basic list of solid Tools offerings here in order to give developers a better sense of what Tools is. I'm ok with updating the list regularly if needed to keep it current. Alternatively, we could link from here to another page that has a more thorough list of available packages. I've talked with Konstantinos about such a list. Related tickets are PLT-444 and PLT-644. 

Are there 2 or 3 more examples we could add here? From the list of a couple of dozen folders on the plutus-apps repo, which ones, if any, do you think would be helpful to reference specifically in this intro? 

This page seems interesting. Would it make sense to mention some of what it mentions? 

https://github.com/input-output-hk/plutus-apps/blob/main/plutus-use-cases/ARCHITECTURE.adoc

Reviewer 3: I was suggesting you remove that line. Currently we have: PAB and chain-index. Soon we will also have Marconi, and the PAB and chain-index versions based on Marconi.

Reviewer 3: I think that listing the projects in plutus-tools with a brief explanation is a good idea, here.

> To get started using the Plutus Tools SDK, see the `Plutus starter template repository 

Reviewer 2: I suggest: "See the `Plutus starter template repository <https://github.com/input-output-hk/plutus-starter>`_ for a simple starter project using the Plutus Tools SDK.

Author: Suggestion incorporated.

Plutus Tools SDK User Guide
===============================

Intended Audience
-------------------------------

This user guide is intended for developers who are authoring distributed applications 
("DApps") by using smart contracts on the Cardano blockchain. 

.. note::
   If you are a developer who wants to contribute to the Plutus Tools SDK project, 
   please refer to documentation residing in the `Plutus Tools SDK repository <https://github.com/
   input-output-hk/plutus-apps>`_.

> Plutus Tools SDK

Reviewer 4: I think the Trace Emulator is worth highlighting here as a tool. It's used for testing plutus contracts on an emulated blockchain. https://github.com/input-output-hk/plutus-apps/blob/main/plutus-contract/src/Plutus/Trace/Emulator.hs

Plutus Tools SDK User Guide
===============================

Intended Audience
-------------------------------

This user guide is intended for developers who are authoring distributed applications 
("DApps") by using smart contracts on the Cardano blockchain. 

.. note::
   If you are a developer who wants to contribute to the Plutus Tools SDK project, 
   please refer to documentation residing in the `Plutus Tools SDK repository <https://github.com/
   input-output-hk/plutus-apps>`_.

Plutus Tools SDK
===============================

The Plutus Tools SDK is a collection of off-chain infrastructure resources built 
for external developers. The `Plutus Tools SDK repository <https://github.com/
input-output-hk/plutus-apps>`_ supports the underlying resources that developers 
need who are writing full applications using Plutus in Haskell, including off-chain 
code. The term “off-chain code” refers to the part of a contract application’s code 

> which runs outside of the blockchain. Off-chain code responds to events happening 

Reviewer 5: Not entirely sure but wasn't it the *on-chain* code that responded to events from off-chain? At least that's what I vaguely remember from the Plutus Pioneer Program.

Reviewer 4: The onchain checks whether certain conditions are met with limited scope. The only "response" it can perform is to invalidate a transaction by returning an error. Off-chain code can respond to any event happening, including on this blockchain or from any other interface.

> Plutus Tools Developer Guide

Author: After a conversation with Konstantinos and Andrew, we're trying on "Plutus Tools SDK User Guide."

===============================

> Plutus Tools are a collection of off-chain infrastructure resources. Plutus Tools are built 

Author: Good point. My first impulse was to use a singular reference, but then I thought about it too much and went with plural!

> for developers who are authoring web-based applications that implement smart contracts on the 

Reviewer 3: ```suggestion
for developers who are authoring applications that implement smart contracts on the 
```

Reviewer 3: There is nothing web-specific about the tools.

Author: Great to know, thanks!

Author: I'm thinking we could say "for developers who are authoring distributed applications that implement smart contracts on the ..." Or are they not necessarily always going to be distributed applications?

> for developers who are authoring web-based applications that implement smart contracts on the 

Reviewer 6: I think cloud-based may be a more accurate term.
https://www.techtarget.com/searchcloudcomputing/definition/cloud-application#:~:text=A%20key%20difference%20between%20cloud,a%20local%20computer%20or%20workstation.

Cardano blockchain. 

> The Plutus Tools Team supports the underlying resources that developers need who are 

Reviewer 6: Is it better to refer to the contents of the repo rather than the team?

Author: Yes, probably so. "The Plutus Tools repository supports the underlying resources..."

writing full applications using Plutus in Haskell, including off-chain code. 

> The term “off-chain code” refers to the part of a contract application’s code which 

Reviewer: Should this just link to the glossary?

Author: Good point. However, I want the reader to have a sufficient general understanding of this concept without needing to leave the landing page.

> runs off the chain, usually as a contract application. 

Reviewer 6: redundant use of contract application

Author: True, did some pruning.

Because there are so many potential use cases for what you may choose to create, you are not 
confined to only one prescribed way of using Plutus Tools. Instead, you can access 
the entire tool box where you can find a collection of utilities and components. Pick and choose 
from the tools and examples that best suit your use case. 

> The Plutus Platform

Reviewer 3: I am not sure about the formulation of this section. It sounds like we have a platform, but I don't think that we really do.

For example I do not know what the Plutus Application Framework consists of, I am not sure we have any end-to-end use cases, etc.

Author: Thanks for your feedback. Do you think the info on this page is still largely accurate: 
https://plutus-apps.readthedocs.io/en/latest/plutus/explanations/platform.html

I think it may be important and useful to have a conversation about how we describe and characterize what Plutus Tools is. I have found this to be a challenging subject to have clarity about. 

I'd also like to learn more about the meaning of the content on these pages: 

https://plutus-apps.readthedocs.io/en/latest/reference/examples.html

https://github.com/input-output-hk/plutus-apps/blob/main/plutus-use-cases/ARCHITECTURE.adoc

Reviewer: no, the stuff about the plutus platform is absolutely not accurate

-------------------------------

The Plutus Tools Repo (https://github.com/input-output-hk/plutus-apps) contains 
the Plutus Platform, which consists of these components: 

* Libraries which implement the Plutus Application Framework
* A selection of end-to-end use cases written with the Plutus Application Framework

> * The Plutus Playground, a web-based playground for learning and writing basic Plutus Applications

Reviewer 3: Didn't we decide we don't want to support this?

Author: I got these points from https://github.com/input-output-hk/plutus-apps#readme. Maybe this info is out of date. Are you referring to the Plutus Playground, or all 3 points here?

Reviewer 3: @asutherlandus are we still supporting the playground?

Author: I'm removing this reference to the Plutus Playground.

> * The Plutus Playground, a web-based playground for learning and writing basic Plutus Applications

Reviewer 6: We should mark Plutus playground deprecated

Author: I'll remove this reference to Plutus Playground. We'll also need to update the readme for https://github.com/input-output-hk/plutus-apps.
